src/download.c

```diff
@@ -26,325 +26,154 @@
-			if (count != 1) {
-				err = -1;
-			}
-			break;
-		}
-
-		c = strchr(buffer, '\n');
-		if (c) {
-			*c = 0;
-		}
-		if (c && (c != buffer) && (*(c - 1) == '/')) {
-			/* strip trailing '/' from directory tar */
-			*(c - 1) = 0;
-		}
-		if (strcmp(buffer, file->hash) != 0) {
-			err = -1;
-			break;
-		}
-		count++;
-	}
-
-	pclose(tar);
-free_tarcommand:
-	free_string(&tarcommand);
-
-	return err;
-}
-
-/* This function will break if the same HASH.tar full file is downloaded
- * multiple times in parallel. */
-int untar_full_download(void *data)
+void *swupd_curl_parallel_download_start(size_t max_xfer, size_t max_xfer_bottom, swupd_curl_download_cb success_cb)
 {
-	struct file *file = data;
-	char *tarfile;
-	char *tar_dotfile;
-	char *targetfile;
-	struct stat stat;
-	int err;
-
-	string_or_die(&tar_dotfile, "%s/download/.%s.tar", state_dir, file->hash);
-	string_or_die(&tarfile, "%s/download/%s.tar", state_dir, file->hash);
-	string_or_die(&targetfile, "%s/staged/%s", state_dir, file->hash);
-
-	/* If valid target file already exists, we're done.
-	 * NOTE: this should NEVER happen given the checking that happens
-	 *       ahead of queueing a download.  But... */
-	if (lstat(targetfile, &stat) == 0) {
-		if (verify_file(file, targetfile)) {
-			unlink(tar_dotfile);
-			unlink(tarfile);
-			free_string(&tar_dotfile);
-			free_string(&tarfile);
-			free_string(&targetfile);
-			return 0;
-		} else {
-			unlink(tarfile);
-			unlink(targetfile);
-		}
-	} else if (lstat(tarfile, &stat) == 0) {
-		/* remove tar file from possible past failure */
-		unlink(tarfile);
-	}
+	struct swupd_curl_parallel_handle *h = calloc(1, sizeof(struct swupd_curl_parallel_handle));
 
-	err = rename(tar_dotfile, tarfile);
-	if (err) {
-		free_string(&tar_dotfile);
-		goto exit;
+	h->mcurl = curl_multi_init();
+	if (h->mcurl == NULL) {
+		goto error;
 	}
-	free_string(&tar_dotfile);
 
-	err = check_tarfile_content(file, tarfile);
-	if (err) {
-		goto exit;
-	}
+	curl_multi_setopt(h->mcurl, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX | CURLPIPE_HTTP1);
 
-	/* modern tar will automatically determine the compression type used */
-	char *outputdir;
-	string_or_die(&outputdir, "%s/staged", state_dir);
-	err = extract_to(tarfile, outputdir);
-	free_string(&outputdir);
-	if (err) {
-		fprintf(stderr, "ignoring tar extract failure for fullfile %s.tar (ret %d)\n",
-			file->hash, err);
-		goto exit;
-		/* TODO: respond to ARCHIVE_RETRY error codes
-		 * libarchive returns ARCHIVE_RETRY when tar extraction fails but the
-		 * operation is retry-able. We need to determine if it is worth our time
-		 * to retry in these situations. */
-	} else {
-		/* Only unlink when tar succeeded, so we can examine the tar file
-		 * in the failure case. */
-		unlink(tarfile);
-	}
+	h->max_xfer = max_xfer;
+	h->max_xfer_bottom = max_xfer_bottom;
+	h->success_cb = success_cb;
+	h->curl_hashmap = hashmap_new(SWUPD_CURL_HASH_BUCKETS, file_hash_cmp, file_hash_value);
```

Comment:
hashmap_new can return NULL if the calloc fails. Are you sure all code that uses h->curl_hashmap after this point is checking for this case?

(copied from out-of-date comment)

Reply:
I'll add one abort() after the calloc(). It looks like that's the approach we're using on swupd right now. Should we have a calloc_or_die() too? Or this is a bit too much?

Reply:
We can add a calloc_or_die() when we are prepared to move all our malloc/callocs over to using it.